### PR TITLE
feat(rust): adapt to zbus 5

### PR DESCRIPTION
**Pull request by imobachgs**
## The main goal

Although [zbus 5](https://github.com/dbus2/zbus/releases/tag/zbus-5.0.0) 5 was released a few days ago, Agama is still using zbus 3. The goal of this PR is to adapt the code to use the latest version.

However, this task is not trivial: version 4 already introduced a good share of breaking changes in zbus API.

## `downcast_ref` returns `Result`

The `downcast_ref` function now returns a `Result` instead of an `Option` with, IMHO, it is the right thing to do. This change has a big impact in our code to interact with NetworkManager, so I took the opportunity to:

* Distinguish between a problem and a missing value (we always returned `None` when something went wrong).
* Use `get_property` and `get_optional_property` to simplify our code a bit.

## Other changes

- Do some refactoring and organization of D-Bus proxies.
- Improve the `network::nm::dbus` module. Let's use `get_property` and `get_optional_property` for better readability and error handling. Please, check [this commit (WIP)](https://github.com/agama-project/agama/commit/c21eda8391e8e21926e36b9b15b3494806320c09) if you are interested in further improvements.
- Drop some unused code.

## Tasks

- [ ] Testing, testing and testing.
- [x] Update some outdated D-Bus definitions.
- [x] Improve the `network::nm::dbus` module.

### CODE REVIEW

**Review by teclator — APPROVED** [review]
LGTM

**Review by jcronenberg — COMMENTED** [review]
I mostly looked at the NM dbus code, because that's what I know, but it looks like a nice improvement all around :slightly_smiling_face:

File: rust/agama-lib/src/network/proxies.rs [hunk 1]
```diff
@@ -21,9 +21,9 @@
 //! D-Bus interface proxies for: `org.opensuse.Agama*.**.*`
 //!
 //! This code was generated by `zbus-xmlgen` `3.1.0` from DBus introspection data.`.
```
Comment (jreidinger) [inline rust/agama-lib/src/network/proxies.rs:23]: NP: I am not sure if we should keep this as it is confusing...I hit it often in past that I have newer zbus_xmlgen installed and result is wrong and need adaptation.

File: rust/agama-lib/src/product/client.rs [hunk 2]
```diff
@@ -123,9 +123,10 @@ impl<'a> ProductClient<'a> {
 
     /// register product
     pub async fn register(&self, code: &str, email: &str) -> Result<(u32, String), ServiceError> {
-        let mut options: HashMap<&str, zbus::zvariant::Value> = HashMap::new();
+        let mut options: HashMap<&str, &zbus::zvariant::Value> = HashMap::new();
+        let value = zbus::zvariant::Value::from(email.to_string());
```
Comment (jreidinger) [inline rust/agama-lib/src/product/client.rs:127]: how long  the email lives? Str has From for &str - https://docs.rs/zvariant/latest/zvariant/struct.Str.html#impl-From%3C%26str%3E-for-Str%3C'a%3E

File: rust/agama-lib/src/network/proxies.rs (hunk 1 above)
Comment (imobachgs) [reply rust/agama-lib/src/network/proxies.rs:23]: That's the intention of separating many of those files into smaller ones, to autogenerate them someday (not yet). In any case, that message is wrong, as we are using version 5 now.

File: rust/agama-lib/src/product/client.rs (hunk 2 above)
Comment (imobachgs) [reply rust/agama-lib/src/product/client.rs:127]: Yes, we do not need to do the conversion.

File: rust/agama-lib/src/storage/client/dasd.rs [hunk 3]
```diff
@@ -21,39 +21,47 @@
...
             object_manager_proxy,
+            introspectable_proxy,
         })
     }
 
     pub async fn supported(&self) -> Result<bool, ServiceError> {
-        let introspect = self.manager_proxy.introspect().await?;
+        let introspect = self.introspectable_proxy.introspect().await?;
         // simply check if introspection contain given interface
         Ok(introspect.contains("org.opensuse.Agama.Storage1.DASD.Manager"))
```
Comment (jreidinger) [inline rust/agama-lib/src/storage/client/dasd.rs:66]: finally they split it to own proxy. Maybe code can be improved even check...but looks like they still have xml parsing separately. So no easy way to improve it.

File: rust/agama-lib/src/storage/client/zfcp.rs [hunk 4]
```diff
@@ -23,45 +23,55 @@
...
             object_manager_proxy,
-            connection,
+            introspectable_proxy,
         })
     }
 
     pub async fn supported(&self) -> Result<bool, ServiceError> {
-        let introspect = self.manager_proxy.introspect().await?;
-        // simply check if introspection contain given interface
+        let introspect = self.introspectable_proxy.introspect().await?; // simply check if introspection contain given interface
```
Comment (jreidinger) [inline rust/agama-lib/src/storage/client/zfcp.rs:74]: hmm, this move of comment is wrong. It is related to the next line, not this one.

File: rust/agama-server/src/dbus.rs [hunk 5]
```diff
@@ -108,18 +109,19 @@ impl DBusObjectChangesStream {
...
     /// * interface: name of the interface to watch.
     fn handle_properties_changed(
-        message: Result<Arc<Message>, zbus::Error>,
+        message: Result<Message, zbus::Error>,
         interface: &str,
     ) -> Option<DBusObjectChange> {
         let Ok(message) = message else {
             return None;
         };
+        // let path = message.header().path()?.to_string();
```
Comment (jreidinger) [inline rust/agama-server/src/dbus.rs:118]: dead code?

File: rust/agama-server/src/network/nm/error.rs [hunk 6]
```diff
@@ -19,7 +19,10 @@
 // find current contact information at www.suse.com.
 
 //! NetworkManager error types
+use std::net::AddrParseError;
+
```
Comment (jreidinger) [inline rust/agama-server/src/network/nm/error.rs:23]: is there reason for this empty line? I do not get why the first one is separated from rest.

File: rust/agama-server/src/network/nm/dbus.rs [hunk 7]
```diff
@@ -834,139 +788,128 @@ fn ip_config_from_dbus(conn: &OwnedNestedHash) -> Option<IpConfig> {
...
+
+    if let Ok(bssid) = get_property::<zvariant::Array>(wireless, "bssid") {
         let bssid: Vec<u8> = bssid
-            .get()
             .iter()
-            .map(|u| *u.downcast_ref::<u8>().unwrap())
-            .collect();
+            .map(|u| u.downcast_ref::<u8>())
+            .collect::<Result<Vec<u8>, _>>()?;
+        // FIMXE: properly handle the failing case
```
Comment (jreidinger) [inline rust/agama-server/src/network/nm/dbus.rs:884]: typo FIMXE ( it kills grep attempt for fixmes )

**Review by jreidinger — APPROVED** [review]
nothing critical

File: rust/agama-lib/src/network/proxies.rs (hunk 1 above)
Comment (imobachgs) [reply rust/agama-lib/src/network/proxies.rs:23]: Actually this one was generated with that version. I will update all those proxies (which are working) on a separate PR if needed.

File: rust/agama-server/src/network/nm/dbus.rs [hunk 8]
```diff
@@ -1293,24 +1216,30 @@ mod test {
         ])];
 
         let ipv4_section = HashMap::from([
-            ("method".to_string(), Value::new("auto").to_owned()),
+            ("method".to_string(), Value::new("auto").try_to_owned()?),
```
Comment (mvidner) [inline rust/agama-server/src/network/nm/dbus.rs:1219]: @imobachgs I know this is just test code, but it bothered me  how the meaningful text was lost among the boilerplate, so I made a helper for readability:

```diff
diff --git a/rust/agama-server/src/network/nm/dbus.rs b/rust/agama-server/src/network/nm/dbus.rs
index 3e4ab1bf6..99e69ad2e 100644
--- a/rust/agama-server/src/network/nm/dbus.rs
+++ b/rust/agama-server/src/network/nm/dbus.rs
@@ -1195,6 +1195,14 @@ mod test {
     use uuid::Uuid;
     use zbus::zvariant::{self, Array, Dict, OwnedValue, Value};
 
+    // hash item
+    fn hi<'a, T>(key: &str, value: T) -> anyhow::Result<(String, OwnedValue)>
+    where
+        T: Into<Value<'a>> + zbus::zvariant::Type,
+    {
+        Ok((key.to_string(), Value::new(value).try_to_owned()?))
+    }
+
     #[test]
     fn test_connection_from_dbus() -> anyhow::Result<()> {
         let uuid = Uuid::new_v4().to_string();
@@ -1216,31 +1231,13 @@ mod test {
         ])];
 
         let ipv4_section = HashMap::from([
-            ("method".to_string(), Value::new("auto").try_to_owned()?),
-            (
-                "address-data".to_string(),
-                Value::new(address_v4_data).try_to_owned()?,
-            ),
-            (
-                "gateway".to_string(),
-                Value::new("192.168.0.1").try_to_owned()?,
-            ),
-            (
-                "dns-data".to_string(),
-                Value::new(vec!["192.168.0.2"]).try_to_owned()?,
-            ),
-            (
-                "dns-search".to_string(),
-                Value::new(vec!["suse.com", "example.com"]).try_to_owned()?,
-            ),
-            (
-                "ignore-auto-dns".to_string(),
-                Value::new(true).try_to_owned()?,
-            ),
-            (
-                "route-data".to_string(),
-                Value::new(route_v4_data).try_to_owned()?,
-            ),
+            hi("method", "auto")?,
+            hi("address-data", address_v4_data)?,
+            hi("gateway", "192.168.0.1")?,
+            hi("dns-data", vec!["192.168.0.2"])?,
+            hi("dns-search", vec!["suse.com", "example.com"])?,
+            hi("ignore-auto-dns", true)?,
+            hi("route-data", route_v4_data)?,
         ]);
 
         let address_v6_data = vec![HashMap::from([
```

**Pull request merged**